lokimq/worker.cpp

```diff
@@ -5,27 +5,93 @@
+        // for an all-clear signal from LokiMQ first, then we fire our `start` callback, then we can
+        // start waiting for commands in the main loop further down.  (We also can't get the
+        // reference to our `tagged_workers` element until the main proxy threads is running).
+
+        waiting_for_command = true;
+
+        while (true) {
+            LMQ_TRACE("tagged worker ", worker_id, " waiting for startup signal");
+            recv_message_parts(sock, parts);
+            if (parts.size() != 1) {
+                LMQ_LOG(error, "Internal error: worker ", worker_id, " received invalid ", parts.size(), "-part startup msg");
+                continue;
+            }
+            auto command = view(parts[0]);
+            if (command == "START"sv) {
+                LMQ_LOG(debug, "Tagged worker ", worker_id, " received start signal, starting up");
+                if (start) start();
+                break;
+            } else if (command == "QUIT"sv) {
+                LMQ_LOG(debug, "Tagged worker ", worker_id, " shutting down");
+                detail::send_control(sock, "QUITTING");
+                sock.setsockopt<int>(ZMQ_LINGER, 1000);
+                sock.close();
+                return;
+            } else {
+                LMQ_LOG(error, "Internal error: worker ", worker_id, " received invalid command: `", command, "'");
+            }
+        }
+    } else {
+        // Otherwise for a regular worker we can only be started by an active main proxy thread
+        // which will have preloaded our first job so we can start off right away.
+        waiting_for_command = false;
+    }
+
+    // This will always contains the current job, and is guaranteed to never be invalidated.
+    run_info& run = tagged ? std::get<run_info>(tagged_workers[index - 1]) : workers[index];
 
     while (true) {
+        while (waiting_for_command) {
+            LMQ_TRACE("worker ", worker_id, " waiting for command");
+            parts.clear();
+            recv_message_parts(sock, parts);
+
+            if (parts.size() != 1) {
+                LMQ_LOG(error, "Internal error: worker ", worker_id, " received invalid ", parts.size(), "-part worker instruction");
+                continue;
+            }
+            auto command = view(parts[0]);
+            if (command == "RUN"sv) {
+                LMQ_LOG(debug, "worker ", worker_id, " running ", run.is_batch_job ? "batch"s : "command " + run.command);
+                waiting_for_command = false;
+            } else if (command == "QUIT"sv) {
+                LMQ_LOG(debug, "worker ", worker_id, " shutting down");
+                detail::send_control(sock, "QUITTING");
+                sock.setsockopt<int>(ZMQ_LINGER, 1000);
+                sock.close();
+                return;
+            } else {
+                LMQ_LOG(error, "Internal error: worker ", worker_id, " received invalid command: `", command, "'");
+            }
+        }
+
```

Comment:
this looks similar to the block of code 40 lines above it, maybe it could be a little more DRY

Reply:
Fixed.